watch/funcs_test.go

```diff
@@ -185,172 +218,243 @@ func TestKeyPrefixWatch(t *testing.T) {
-				return
+				return errNoContent
 			}
 			v, ok := raw.(map[string][]string)
 			if !ok || v["consul"] == nil {
-				t.Fatalf("Bad: %#v", raw)
+				return errBadContent
 			}
-			invoke++
-		}
+			return nil
+		}()
 	}
 
+	var wg sync.WaitGroup
+	wg.Add(1)
 	go func() {
-		time.Sleep(20 * time.Millisecond)
-		plan.Stop()
+		defer wg.Done()
+		agent := a.Client().Agent()
 
-		agent := plan.client.Agent()
+		time.Sleep(20 * time.Millisecond)
 		reg := &consulapi.AgentServiceRegistration{
 			ID:   "foo",
 			Name: "foo",
 		}
-		agent.ServiceRegister(reg)
+		if err := agent.ServiceRegister(reg); err != nil {
+			t.Fatalf("err: %v", err)
+		}
+
 		time.Sleep(20 * time.Millisecond)
-		agent.ServiceDeregister("foo")
+		if err := agent.ServiceDeregister("foo"); err != nil {
+			t.Fatalf("err: %v", err)
+		}
 	}()
 
-	err := plan.Run(consulAddr)
-	if err != nil {
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		if err := plan.Run(a.HTTPAddr()); err != nil {
+			t.Fatalf("err: %v", err)
+		}
+	}()
+
+	// wait for svc reg
+	if err := <-invoke; err != nil {
 		t.Fatalf("err: %v", err)
 	}
 
-	if invoke == 0 {
-		t.Fatalf("bad: %v", invoke)
+	// wait for svc unreg
+	if err := <-invoke; err != nil {
+		t.Fatalf("err: %v", err)
 	}
+
+	plan.Stop()
+	wg.Wait()
 }
 
 func TestNodesWatch(t *testing.T) {
-	if consulAddr == "" {
-		t.Skip()
-	}
+	a := agent.NewTestAgent(t.Name(), ``)
+	defer a.Shutdown()
+
+	invoke := make(chan error)
 	plan := mustParse(t, `{"type":"nodes"}`)
-	invoke := 0
 	plan.Handler = func(idx uint64, raw interface{}) {
-		if invoke == 0 {
+		invoke <- func() error {
 			if raw == nil {
-				return
+				return errNoContent
 			}
 			v, ok := raw.([]*consulapi.Node)
 			if !ok || len(v) == 0 {
-				t.Fatalf("Bad: %#v", raw)
+				return errBadContent
 			}
-			invoke++
-		}
+			return nil
+		}()
 	}
 
+	var wg sync.WaitGroup
+	wg.Add(1)
 	go func() {
-		time.Sleep(20 * time.Millisecond)
-		plan.Stop()
+		defer wg.Done()
+		catalog := a.Client().Catalog()
 
-		catalog := plan.client.Catalog()
+		time.Sleep(20 * time.Millisecond)
 		reg := &consulapi.CatalogRegistration{
 			Node:       "foobar",
 			Address:    "1.1.1.1",
 			Datacenter: "dc1",
 		}
-		catalog.Register(reg, nil)
+		if _, err := catalog.Register(reg, nil); err != nil {
+			t.Fatalf("err: %v", err)
+		}
+
 		time.Sleep(20 * time.Millisecond)
 		dereg := &consulapi.CatalogDeregistration{
 			Node:       "foobar",
 			Address:    "1.1.1.1",
 			Datacenter: "dc1",
 		}
-		catalog.Deregister(dereg, nil)
+		if _, err := catalog.Deregister(dereg, nil); err != nil {
+			t.Fatalf("err: %v", err)
+		}
 	}()
 
-	err := plan.Run(consulAddr)
-	if err != nil {
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		if err := plan.Run(a.HTTPAddr()); err != nil {
+			t.Fatalf("err: %v", err)
+		}
+	}()
+
+	// wait for kv put
+	if err := <-invoke; err != nil {
 		t.Fatalf("err: %v", err)
 	}
 
-	if invoke == 0 {
-		t.Fatalf("bad: %v", invoke)
+	// wait for kv del
+	if err := <-invoke; err != nil {
```

Comment:
About the `invoke` channel approach here and everywhere else - if there's a bug introduced in the future in watch handler invocation causing the invoke channel to never be written to, the test won't fail it will just hang at this line, and then it will eventually timeout based on the timeout in the makefile. Wondering if there's a more robust way to declare failure - wp.Run should invoke the handler if its working correctly. What if you closed the invoke channel in line 334 above after wp.Run returns?.  If the channel is closed and nothing was received from it we can fail the test immediately.

Reply:
The main reason I've switched to an unbuffered channel is that it provides the sync we need for the multiple go routines at play. The one thing that is missing is the notification from the `Run` method that it is indeed blocking now. That is one reason we still need the timers IMO. You have to wait for `Run` to be in the right state before making the first change otherwise you could miss that. 

Closing the `invoke` channel in 334 won't work since you have to call `plan.Stop()` first for `Run` to terminate. When you do that `Run` may exit immediately and may not handle another watch event depending on where it is in the loop. The previous setup was quite racy IMO.

I am aware that tests will hang (timeout) if something goes wrong and to make it more robust we could have them timeout quicker. However, I've accepted this outcome as a risk for more predictable behavior in the tests. 

Now that I think about this I think we could have another go routine which injects an error after some time (e.g. 2s) 

```go
var timeout = 5 * time.Second

func makeInvokeCh() chan error {
    ch := make(chan error)
    time.AfterFunc(timeout, func() { ch <- errTimeout })
    return ch
}

func TestXXX(t *testing.T) {
    invoke := makeInvokeCh()
    ...    
}
```

Reply:
:+1: this 5 second timeout makes sense. I forgot that Run may exit without having called the handler